### PR TITLE
👻 Spoooky! Ctrl+Z Fix!

### LINKED ISSUE #1103: ctrl-z problem on Linux
Launching a command (like emacs) followed by ctrl-z suspends
the process but doesn't return keyboard control to xonsh. So there
is no way to detach the process with bg and continue typing. The
documentation only mentions that ctrl-z doesn't work on Windows.

xonfig
+------------------+----------+
| xonsh | 0.3.2 |
| Python | 3.4.3 |
| PLY | 3.7 |
| have readline | True |
| prompt toolkit | None |
| shell type | readline |
| pygments | None |
| on posix | True |
| on linux | True |
| distro | Fedora |
| on darwin | False |
| on windows | False |
| is superuser | False |
| default encoding | utf-8 |
+------------------+----------+

### PR DESCRIPTION
**Pull request by scopatz**
This PR fixes the many job control issues we have had with suspending and presumably closes #1103, #2353, #2370, #2874, #3019, and #3216. I am writing this down so that we don't have to rediscover all of these points in the future.

TL;DR is that threadable subprocesses can't work with job suspension in their current form.

## The Problem

Pressing `Ctrl+Z` (`^Z`, or `b"\x1a"`) at the interactive terminal on posix would cause the subprocess to go into state T (or [stopped](https://linuxjourney.com/lesson/process-states)). This is because [`SIGTSTP`](https://www.gnu.org/software/libc/manual/html_node/Job-Control-Signals.html) is being issued (as it should be).  However, this would cause xonsh / Python to deadlock until an *external* process sent a `SIGCONT` signal, **which is very bad**. 

For example, let's look at the `sleep` command run in a thread (`PopenThread`) before `^Z` is pressed. The process status and heirarchy would look like:

**Before `^Z`:**

```
PID   Status Command
10016 S      |- xonsh   # this is the parent xonsh process
10045 S       |- xonsh  # some xonsh i/o monitoring thread
10044 S       |- xonsh  # some xonsh i/o monitoring thread
10043 S       |- xonsh  # some xonsh i/o monitoring thread
10042 S       |- sleep  # the subprocess being run
```

They are all asleep like they should be.  Oddly though, the threads that are spun up to monitor stdout, stderr, etc. each get their own process id and are children of original xonsh process at the same level as the subprocess we are running.

Now suposse that the user hits `^Z`. The status of the processes becomes:

**After `^Z`:**

```
PID   Status Command
10016 S      |- xonsh   # this is the parent xonsh process
10045 S       |- xonsh  # some xonsh i/o monitoring thread
10044 S       |- xonsh  # some xonsh i/o monitoring thread
10043 S       |- xonsh  # some xonsh i/o monitoring thread
10042 T       |- sleep  # the subprocess being run
```

Because these extra xonsh threads exist that are waiting on the stopped process, xonsh itself is deadlocked and cannot act on any further input until the `SIGCONT` signal is recieved. Note that this situation is very different for subprocs that are not threaded (`Popen`).  For example, let's look at `less` running on some file. **These procs behave normally with job control.**

**Before `^Z`:**

```
PID   Status Command
10016 S      |- xonsh        # this is the parent xonsh process
10042 R       |- less f.txt  # the subprocess being run
```

**After `^Z`:**

```
PID   Status Command
10016 S      |- xonsh        # this is the parent xonsh process
10042 T       |- less f.txt  # the subprocess being run
```

The example with `less` parallels what happens in Bash, where no subprocs are threadable.

## Strategies

There are so many options for how to try to resolve the deadlocking, I named them after character(s) in a [book series](https://www.goodreads.com/book/show/34403860-sufficiently-advanced-magic). Branches have the same name as strategy. The stratgies are outlined here, with sections going into their feasability below.

* **Keras:** Kill the extra threads before suspending.
* **Selyrian:** Kill the extra threads and ignore the suspending signal.
* **Talien:** Ignore the suspend signal and try to reimplement it by reading stdin
* **Salaris:** Ignore the suspend signal entriely and just continue on as if nothing happened.
* **Lydia:** The addition of an environment variable to allow the user to determine if subprocs should be run in threads or not.

The above (except for Lydia) apply only to `PopenThread`.

## Keras

The Keras strategy attempts to kill the threads prior to suspending. This means turning off the teeing of stdin, stdout, and stderr. In principle, this is a safe operations because we will be stopping the subprocess and returning control back to the terminal, and then rendering a new prompt. 

However, this turns out to not be possible because the subproc seems to be stopped before the other processes and thus any signal handlers that have been installed by Python/xonsh to address `SIGTSTP` are never called and we are deadlocked. It may be that the signal handlers are never called just because the subprocess *always seems* to have a lower process id (PID) than the other xonsh threads, and thus is stopped first.  However, even if were were able to manipulate the execution order to ensure that the threads have lower PIDs, this seems like a non-detrministic fix.

It is important to note, though, that SIGTSTP comes in on file descriptor 0 (fd0, stdin), which has been attached to the subprocess by this point. So the signal is first sent to the subprocess. We do not want to, and may not be able to, redirect stdin for all subprocs. Many commands rely on being attched to fd0 for stdin, or change their behavior if they think they are being piped a different input.

One way to get around this might have been to run the subprocess in a Python subprocess, adding a level to the Python process heirarchy.  For example:

**Before `^Z`:**

```
PID   Status Command
10016 S      |- xonsh   # this is the parent xonsh process
10046 S       |- xonsh  # some xonsh i/o monitoring thread
10045 S       |- xonsh  # some xonsh i/o monitoring thread
10044 S       |- xonsh  # some xonsh i/o monitoring thread
10042 S       |- python -c "subprocess.run(['sleep', '2'])" 
10043 S        |- sleep 2 # the subprocess being run
```

You might think this is a reasonble cheat because unthreaded (`Popen`) subprocs have the correct job control behavior in xonsh. However, this tomfoolery doesn't trick anybody and the deadlock remains.

**After `^Z`:**

```
PID   Status Command
10016 S      |- xonsh   # this is the parent xonsh process
10046 S       |- xonsh  # some xonsh i/o monitoring thread
10045 S       |- xonsh  # some xonsh i/o monitoring thread
10044 S       |- xonsh  # some xonsh i/o monitoring thread
10042 T       |- python -c "subprocess.run(['sleep', '2'])" 
10043 T        |- sleep 2 # the subprocess being run
```

Thus, unfortunately, the Keras strategy is unviable.

## Selyrian

The Seylrian strategy is to kill the threads and the ignore any suspending that may happen. As discussed with the Keras strategy, we are not able to kill the threads. Thus, unfortunately, the Selyrian strategy is unviable.

## Taelien

The Taelien strategy is to ignore the `^Z`, kill the other threads, and then suspend the subprocess ourselves. The theory behind this is that it is possible to set suspends signal to be bound to a null key (`b"\x00"`, `stty susp "<undef>"`). In this case, when the user types `^Z`, the kernel does not send SIGTSTP to the subprocess. However, because `^Z` is a control character, it is still echoed to stdin/stdout/stderr. It should be possible to read the `^Z` from the relevant buffer `PopenThread` and hook into the proper next steps.

The good news is that unbinding the suspend action in this way does prevent `^Z` from sending SIGTSTP, and the subprocess continues executing as if nothing had happened.  The bad news is that it is not possible to monitor the I/O buffers for the `^Z` (`b"\x1a"`) character because it written directly to that buffer (by the kernel) and does not go through the capture pipeline setup by `SubprocSpec` and `PopenThread`. Also because of this direct writting, we have no way of realizing that `^Z` has been typed and cannot issue a message to the user that it has been disabled.

Thus, unfortunately, the Taelien strategy is unviable.

## Salaris

The Salaris strategy is to just ignore all `^Z` input completely in `PopenThread`. Given the discussion in the Taelien strategy, this...actually works.

## Lydia

The Lydia strategy yields initial control over whether `Popen` or `PopenThread` is selected to the user. This is presumably done through an environment variable (`$THREAD_SUBPROCS`, because no one else in the world seems to be using it). This allows users to choose the behavior they want, both as a global configuration and in a small context with `${...}.swap(THREAD_SUBPROCS=False)`. This is effectively an easier-to-use alternative to saying,

```python
__xonsh__.commands_cache.threadable_predictors['cmd'] = lambda x: False
```

when testing if threading subprocs is causing problems. It also enables people to always have a working version of `^Z` if that is what they want. The trade off is that when `$THREAD_SUBPROCS = False`, the I/O streams may not be fully captured on the command pipeline in all cases.

The Lydia strategy is also workable.

## Solution

In the end, the only two viable strategies are Salaris + Lydia.  This PR implements these both, `lydia`  being branched off of `salaris`. This is not the ideal combination for all of the reasons above.  However, it is the best that is achievable given the current state of the code base.  

As discussed many other times, we can probably do a ground-up rewrite of this whole system and incorporate all of the lessons learned. This is probably best done in another library (slug) that xonsh then uses. However, it is not clear to me that the Python `subprocess` module is actually able to handle the kind of signal manipulation we want. We might need to move to an async system or write a low-level version of job control for proper handling of `^Z`.

### CODE REVIEW

**Comment by astronouth7303** [comment]
Terrible. Try again in 2 months.

**Comment by scopatz** [comment]
What do you mean? Does it not work?

**Comment by scopatz** [comment]
Or am I just too into LitRPG ATM 🤣

**Comment by astronouth7303** [comment]
No, Just the title being the wrong holiday. :P

I'm going to try to read & comprehend this soon. It's a lot.

**Comment by laloch** [comment]
> One way to get around this might have been to run the subprocess in a Python subprocess, adding a level to the Python process heirarchy.

I'd say that this could be a way to go. All you need is to install custom signal handler in the wrapper process. Try
```
python -c "import signal\nimport subprocess\nsignal.signal(signal.SIGTSTP, signal.SIG_IGN)\nsubprocess.run(['sleep', '2'])"
```

**Comment by laloch** [comment]
Slightly more elaborate code of the wrapper process:
```python
import signal
import subprocess

def handle_tstp(s, f):
    # TODO: terminate monitoring threads
    p.communicate(b'\x1a')

signal.signal(signal.SIGTSTP, handle_tstp)
p=subprocess.Popen(['sleep', '2'])
p.wait()
```

**Comment by laloch** [comment]
@scopatz, are you sure about those numbers in "PID" column of yours being PIDs? I'd say they're rather SPIDs. For me the (presumably) same tree looks like:
```
$ ps u -Tg58030
USER         PID    SPID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
laloch     58030   58030  1.1  0.6 2146752 52288 pts/1   Ssl  16:32   0:04 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58030   58474  0.2  0.6 2146752 52288 pts/1   Ssl  16:33   0:00 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58030   58475  0.0  0.6 2146752 52288 pts/1   Ssl  16:33   0:00 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58030   58476  0.0  0.6 2146752 52288 pts/1   Ssl  16:33   0:00 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58473   58473  0.0  0.1  17420  9476 pts/1    T+   16:33   0:00 /usr/bin/python -c import subprocess p=subprocess.Popen(['sleep', '2']) p.wait()
laloch     58477   58477  0.0  0.0   8448   692 pts/1    T+   16:33   0:00 sleep 2
```
The three queue threads have the same PID, but different SPIDs, which is just expected.
Using the above python wrapper process the situation after ^Z looks like this:
```
$ ps u -Tg58030
USER         PID    SPID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
laloch     58030   58030  0.9  0.6 2138812 52616 pts/1   Ssl  16:32   0:05 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58030   58947  1.0  0.6 2138812 52616 pts/1   Ssl  16:41   0:00 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58030   58948  0.0  0.6 2138812 52616 pts/1   Ssl  16:41   0:00 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58030   58949  0.0  0.6 2138812 52616 pts/1   Ssl  16:41   0:00 /usr/bin/python -u /usr/local/bin/xonsh
laloch     58946   58946  0.2  0.1  17468  9592 pts/1    S+   16:41   0:00 /usr/bin/python wrap.py
laloch     58950   58950  0.0  0.0   8448   756 pts/1    T+   16:41   0:00 sleep 2
```

**Comment by scopatz** [comment]
> No, Just the title being the wrong holiday. :P

Ahahah. Ok fixed!

> I'm going to try to read & comprehend this soon. It's a lot.

Yeah it is a lot.  I spent basically all of my free time last week (while my partner was out of town) just trying to figure this out. I had started down this road so many times before that I didn't want to lose what I had learned again.

> All you need is to install custom signal handler in the wrapper process.

So yeah that might work, I hadn't tried that specifically

> # TODO: terminate monitoring threads

So that is the wrong place to terminate the monitoring threads.  It is the parent that had the monitoring threads.  Also, keep in mind that one of the monitoring threads is basically doing `os.read(1, 1024)` (i.e. 1 KB from stdout) and it is this that is deadlocking.

>  I'd say they're rather SPIDs

Yeah, I am using  `htop` to monitor them, so they are probably SPIDs, rather than PIDs.  But `htop` displays them as "PID"

**Comment by laloch** [comment]
> So that is the wrong place to terminate the monitoring threads. It is the parent that had the monitoring threads.

You might be right about this being the wrong place, but you can IMO `os.kill()` them from within the wrapper anyway, once you know their SPIDs, which is non-trivial by itself, because `threading.Thread.ident` is not native thread ID (WTF?). Once you manage to `os.kill()` them, the parent process raises (some kind of) I/O error exception. You are free to handle the exception and dispose the killed queue threads gracefully.

**Comment by scopatz** [comment]
Yeah, but you need some sort of nanny process to do the `os.kill()`, I think.

**Comment by scopatz** [comment]
OK! I tried your more sophisticated example @laloch on the `keras` branch.  It *does* allow more inteaction from the use. Specifically `^C` will actually kill the python subprocess, but leave the sub-subprocess (e.g. `sleep`) in a zombie (`Z`) state.  It also lets us inject a message to the user in the python subprocess signal handler.

However, we still aren't able to kill the I/O reading threads in the top-level xonsh process, as Popen has replaced all signal handlers. Thus we never get back to the job control we want. Is there a way for the python subprocess to return without killing the sub-subprocess and yield control back to xonsh?

**Comment by laloch** [comment]
> Specifically ^C will actually kill the python subprocess, but leave the sub-subprocess (e.g. sleep) in a zombie (Z) state.

Try to replace `p.communicate(b'\x1a')` by `p.send_signal(signal.SIGSTOP)`. `communicate` blocks for some reason in the signal handler.

**Comment by laloch** [comment]
Just for the record, there is pure python terminal multiplexer supporting Linux and Windows: https://github.com/prompt-toolkit/pymux

**Comment by scopatz** [comment]
Ok! I just tried `p.send_signal(signal.SIGSTOP)` and it is the same as communicate. ^C and other things work, and we can print a message, but there is no job control back in the top-level xonsh process. 

I wonder how much of the multiplexing we can use as a library...

File: xonsh/jobs.py [hunk 1]
```diff
@@ -202,7 +202,7 @@ def wait_for_active_job(last_task=None, backgrounded=False, return_error=False):
                     last_task=active_task, backgrounded=backgrounded
                 )
         if os.WIFSTOPPED(wcode):
-            print("^Z")
+            print()
```
[the remark this answers is not in the document]
Comment (scopatz) [reply xonsh/jobs.py]: I think it could be removed potentially.

**Comment by scopatz** [comment]
Thanks for the review @gforsyth! I have fixed this up! I look forward to hearing more suggestions for improvement here

**Comment by scopatz** [comment]
Pinging this again, since it has been a few days.  Basically the advantage here is that it lets users turn off PopenThread and doesn't deadlock them if it is on.

**Comment by gforsyth** [comment]
Hey @scopatz -- I think we can put this in, even to just remediate the deadlocking issues.

One initial hiccup while testing this: If I open up `vim` and Ctrl-Z, it correctly drops me back to the shell prompt, but running `bg` then foregrounds the job instead.

```
(base)   ~ $ $THREAD_SUBPROCS=False
(base)   ~ $ vim hithere
(base)   ~ $ bg
[1]+ running: vim hithere (67932)
(base)   ~ $ fg
[1]+ running: vim hithere (67932)
(base)   ~ $ bg
[1]+ running: vim hithere (67932)
(base)   ~ $ bg 1
[1]+ running: vim hithere (67932)
(base)   ~ $ fg 1
[1]+ running: vim hithere (67932)
```

Also totally cool with me if we tackle this in a separate PR.

**Comment by scopatz** [comment]
> One initial hiccup while testing this: If I open up vim and Ctrl-Z, it correctly drops me back to the shell prompt, but running bg then foregrounds the job instead.

Ahh interesting. Thanks for bringing this up.  This was probably a latent bug in job control. I'd prefer if we did that in a separate PR / issued a separate bug report for it.

**Comment by melund** [comment]
I am ok with this. Let's just get it to in :)

**Comment by scopatz** [comment]
Thanks All!

**Comment by con-f-use** [comment]
This might have broken a few things: 
 - https://github.com/direnv/direnv/issues/527

        $ b = !(echo test)          # shouldn't have output                                                                                                                                                                                                                                                  
        test
        $ b                                                                                                                                                                                                                                                                            
        CommandPipeline(stdin=None, stdout=None, stderr=None, pid=13201, returncode=0, args=['echo', 'test'], alias=None, stdin_redirect=['<stdin>', 'r'], stdout_redirect=['<stdout>', 'a'], stderr_redirect=['<stderr>', 'r'], timestamps=[1571826435.7078345, 1571826438.2024937], executed_cmd=['/bin/echo', 'test'], input=, output=, errors=None)
        $ b.output                  # should have captured output                                                                                                                                                                                                                                                      
        ''
 - https://github.com/xonsh/xonsh/issues/3366, https://github.com/direnv/direnv/issues/528

        # using commit 5648b8d
        $ echo "hello" | @(lambda a, s=None: s.read().strip() + " world\n")            
        ^C^C^C^C^C^X   # Waiting forever, cannot abort with Ctrl + C

**Comment by anki-code** [comment]
Epic consequences ;)
